p-ata/src/entrypoint.rs

```diff
@@ -0,0 +1,126 @@
+/// ```
+///
+/// ## Account Layout (Create)
+/// ```ignore
+/// [0] payer                    (signer, writable) - pays for account creation (and rent if applicable)
+/// [1] associated_token_account (writable)         - account to create  
+/// [2] wallet                   (signer)           - token account owner
+/// [3] mint                                        - token mint account
+/// [4] system_program                              - system program
+/// [5] token_program                               - token program
+/// [6] rent_sysvar              (optional)         - rent sysvar (for optimization)
+/// ```
+///
+/// ## Security
+///
+/// - All bump hints are validated for canonicality
+/// - `token_account_len` is bounded by `MAX_SANE_ACCOUNT_LENGTH`
+#[inline(always)]
+pub fn process_instruction(
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    data: &[u8],
+) -> ProgramResult {
+    match data {
+        // Empty data defaults to Create (discriminator 0) - preserving backward compatibility
+        [] => process_create_associated_token_account(program_id, accounts, false, None, None),
+        [discriminator, instruction_data @ ..] => {
+            let idempotent = match *discriminator {
+                0 => false,
+                1 => true,
+                2 => {
+                    return match instruction_data {
+                        [] => process_recover_nested(program_id, accounts, None),
+                        [owner_bump, nested_bump, destination_bump] => process_recover_nested(
+                            program_id,
+                            accounts,
+                            Some((*owner_bump, *nested_bump, *destination_bump)),
+                        ),
+                        _ => Err(pinocchio::program_error::ProgramError::InvalidInstructionData),
+                    }
+                }
+                _ => return Err(pinocchio::program_error::ProgramError::InvalidInstructionData),
+            };
+
+            match instruction_data {
+                // No additional data - compute bump and account_len on-chain (original behavior)
+                [] => process_create_associated_token_account(
+                    program_id, accounts, idempotent, None, None,
+                ),
+                // Only bump provided
+                [bump] => process_create_associated_token_account(
+                    program_id,
+                    accounts,
+                    idempotent,
+                    Some(*bump),
+                    None,
+                ),
+                // Bump + account_len provided (for Token-2022 optimization)
+                [bump, account_len_bytes @ ..] => {
+                    // SAFETY: runtime-bounded, and account_len is last.
+                    let account_len = unsafe {
+                        u16::from_le_bytes(*(account_len_bytes.as_ptr() as *const [u8; 2]))
+                    };
```

Comment:
This needs to check that there's at least two bytes, right? We shouldn't read past the provided instruction data